feat: implement configurable push notifications system

Sarthak-btr opened the pull request:
Implement Push Notifications System using Firebase Cloud Messaging (FCM)

### Description

This PR implements a comprehensive push notifications system using Firebase Cloud Messaging (FCM) for the React Native boilerplate application. The implementation provides a configurable notification framework that supports both iOS and Android platforms with environment-variable-based configuration for enhanced security and flexibility across different deployment environments.

### Key Features

* **Environment-variable-based Firebase configuration** (replacing traditional `google-services.json` and `GoogleService-Info.plist` files)
* **Foreground and background notification handling**
* **Configurable notification preferences** with local storage persistence
* **Integration with existing app architecture through React contexts**

The system uses a service-oriented architecture with `FirebaseMessagingService` handling FCM operations and `NotificationService` managing local notification display and scheduling.

### Database Schema Changes

No database schema changes were made as part of this PR.

### Tests

#### Automated Test Cases Added

* Added Jest mocks for `@react-native-firebase/app` module to prevent CI/CD pipeline failures
* Added Jest mocks for `@react-native-firebase/messaging` module with comprehensive method mocking

#### Manual Test Cases Run

**Firebase Configuration Test:**

1. Configure environment variables in `.env` file with valid Firebase project credentials
3. Verify Firebase initialization logs appear in console
4. Confirm no Firebase configuration errors occur during app startup

**Push Notification Permission Request:**

1. Install the app on a fresh device/simulator
2. Launch the app for the first time
4. Check console logs for FCM token generation
5. Confirm permission status is correctly logged

**Foreground Notification Handling:**

1. Send a test push notification while app is in foreground using Firebase Console
2. Verify notification is received and logged in console
3. Confirm notification display handler is triggered
4. Test with custom notification payload data

**Background Notification Handling:**

1. Send a test push notification while app is in background
2. Verify notification appears in system notification tray
3. Tap notification to open app
4. Confirm notification opened handler is triggered and logged

**FCM Token Management:**

1. Launch app and verify FCM token is generated and logged
2. Test token refresh scenarios by clearing app data and relaunching
3. Verify new tokens are generated and properly logged
4. Confirm token persistence across app sessions

Related to #51

Sarthak-btr committed:
feat: enhance Change API URL button with FAB design

- Replace corner-positioned gear icon with prominent FAB
- Add circular container with white background and primary border
- Improve accessibility with proper labels and hints
- Create dedicated styles hook for theme integration
- Better positioning (bottom: -10, right: -20) for visibility
- Use primary color scheme for better visual hierarchy

Resolves: Improves discoverability of API URL change feature in dev environments

Sarthak-btr committed:
feat: add NotificationService for handling push notifications with Firebase and react-native-push-notification

- Implemented NotificationService class to manage push notifications.
- Integrated Firebase Messaging for background message handling.
- Configured react-native-push-notification for local notifications.
- Created a default notification channel for Android.
- Added methods for displaying background notifications and handling foreground notifications.

Sarthak-btr committed:
feat: Implement Firebase Messaging Service and Notification Setup Hook

- Added FirebaseMessagingService to handle FCM token retrieval, background and foreground message handling, and notification management.
- Created useNotificationSetup hook to initialize Firebase messaging and Notifee notification service, including setup for foreground event handlers and cleanup on unmount.

Sarthak-btr committed:
refactor: Remove unused styles and simplify ChangeApiUrlButton component

Sarthak-btr committed:
feat: Add notification context and types for managing user notification preferences

Sarthak-btr committed:
feat: Integrate NotificationContextProvider and simplify notification setup hook

Sarthak-btr committed:
refactor: Remove console log statements for cleaner notification service code

Sarthak-btr committed:
feat: Implement Firebase messaging and notification handling in AppDelegate

Sarthak-btr committed:
chore: Bump version to 0.0.3 in package.json

Sarthak-btr committed:
Merge branch 'main' into sarthak/feat/implement-push-notifications-support

Sarthak-btr committed:
test: Add Jest mocks for Notifee notification handling

jjalan commented on src/contexts/notification-context.tsx:
```diff
@@ -0,0 +1,139 @@
+  const [settings, setSettings] = useState<NotificationSettings>(DEFAULT_NOTIFICATION_SETTINGS);
+  const [isLoading, setIsLoading] = useState(true);
+
+  // Load saved settings on mount
```
Avoid comments like this, here and other places. See our guidance on comments:

[How to write better code comments](https://dev.to/adammc331/todo-write-a-better-comment-4c8c)

Sarthak5290 replied:
updated

jjalan commented on ios/Boilerplate/AppDelegate.h:
```diff
@@ -1,6 +1,95 @@
+    center.delegate = self;
+
+    // Call the parent implementation
+    self.moduleName = @"YourAppName"; // Replace with your actual app name
```
This probably needs to be set to our brand name?

Sarthak5290 replied:
updated

jjalan commented on android/app/google-services.json:
```diff
@@ -0,0 +1,29 @@
+{
+  "project_info": {
+    "project_number": "752081205061",
```
We would need to update README on how notification infra is setup and which keys needs to be changed and where.

Sarthak-btr replied:
updated

jjalan commented on android/app/google-services.json:
```diff
@@ -0,0 +1,29 @@
+{
+  "project_info": {
+    "project_number": "752081205061",
+    "project_id": "notifications-demo-app-cb35f",
```
The name seems incorrect.

Sarthak5290 replied:
updated

jjalan commented on android/app/google-services.json:
```diff
@@ -0,0 +1,29 @@
+  "project_info": {
+    "project_number": "752081205061",
+    "project_id": "notifications-demo-app-cb35f",
+    "storage_bucket": "notifications-demo-app-cb35f.firebasestorage.app"
```
The name seems incorrect.

Sarthak5290 replied:
updated

jjalan commented on src/types/notification.types.ts:
```diff
@@ -0,0 +1,27 @@
+export type NotificationType =
```
Can you give the background how did we arrive on these types?

Sarthak5290 replied:
in issue https://github.com/jalantechnologies/react-native-template/issues/51

Acceptance Criteria:
Notifications should be configurable (e.g., types, sounds, etc.).

But did not mention anything else

jjalan commented on src/types/notification.types.ts:
```diff
@@ -0,0 +1,27 @@
+  type: NotificationType;
+  title: string;
+  body: string;
+  data?: any;
```
Is it possible to make `NotificationPayload` templatized so instead of `any`, it becomes the type T?

Sarthak-btr replied:
updated

jjalan commented on src/services/notification-service.ts:
```diff
@@ -0,0 +1,273 @@
+    try {
+      await notifee.setBadgeCount(count);
+    } catch (error) {
+      console.error('Error setting badge count:', error);
```
Avoid using console.log. I believe we have logger in the repo

Sarthak-btr replied:
updated

jjalan reviewed (commented):
Can you add screenshots on app behaviour when user denies the permission and system tends to send the notification?

Sarthak-btr committed:
chore: update jest configuration and setup for improved mocking

- Updated jest.config.js to include @react-native-firebase in transformIgnorePatterns.
- Enhanced jest.setup.js with mocks for AsyncStorage, MMKV, react-native-config, react-native-safe-area-context, react-native-gesture-handler, and React Native Firebase modules.
- Updated package.json and yarn.lock to upgrade @react-native-firebase/app and @react-native-firebase/messaging to version 22.2.1.

Sarthak-btr committed:
fix: update google-services.json with correct project configuration and API keys

Sarthak-btr committed:
docs: Update README with Firebase configuration steps for Android and iOS

feat: Enhance notification service with improved error handling and logging

refactor: Simplify notification context state management and improve type definitions

chore: Add GoogleService-Info.plist for Firebase integration

Sarthak-btr committed:
fix: change messaging property to readonly for improved encapsulation

Sarthak-btr committed:
Add Podfile.lock for iOS dependencies with Datadog SDK and React Native updates

Sarthak-btr committed:
feat: integrate Logger for error handling in notification services and improve documentation

jjalan commented on README.md:
````diff
@@ -288,3 +288,65 @@ To fix this, follow these steps:
+- Open your `google-services.json` file.
+- Locate the `package_name` field:
+  ```json
+  "package_name": "com.example.yourapp"
````
How would this work where the identifier changes with build, ex [here](https://github.com/jalantechnologies/react-native-template/pull/148)

Sarthak-btr replied:
updated Readme.md file with :
If the app build identifier changes (e.g., you update the package name or bundle ID), you must create a new application in FCM with the updated identifier (package_name or BUNDLE_ID).

jjalan commented on android/app/google-services.json:
```diff
@@ -0,0 +1,86 @@
+    }
+  ],
+  "configuration_version": "1"
+}
```
add end of line character to avoid `circle with -` icon.

This is true for other files as well

Sarthak5290 replied:
updated

jjalan commented on jest.setup.js:
The changes in this file seems unrelated

Sarthak-btr replied:
restore jest.setup.js configuration

jjalan commented on package.json:
```diff
@@ -21,6 +21,11 @@
     "@datadog/mobile-react-native": "^2.8.0",
     "@datadog/mobile-react-native-session-replay": "^2.8.0",
     "@datadog/mobile-react-navigation": "^2.8.0",
+    "@notifee/react-native": "^9.1.8",
```
We want to avoid taking dependency on third party to display the notification.

Its fine to use firebase to push the data to the app, but our app should do everything after that.

Sarthak-btr replied:
[refactor](https://github.com/jalantechnologies/react-native-template/pull/157/commits/baa5e63090cdb2f0b9ce71c4002cc6ebcb9a189d): remove Notifee integration and replace with react-native-push-notification for notification handling

jjalan commented on package.json:
```diff
@@ -21,6 +21,11 @@
     "@datadog/mobile-react-native-session-replay": "^2.8.0",
     "@datadog/mobile-react-navigation": "^2.8.0",
+    "@notifee/react-native": "^9.1.8",
+    "@react-native-async-storage/async-storage": "^2.2.0",
```
I believe we already have an async storage hook.

Sarthak-btr replied:
[Updated](https://github.com/jalantechnologies/react-native-template/pull/157/commits/76472c6aea60dd618bcf695dd50e5b9e6b6ff423) with the useLocalStorage from utils.

Sarthak-btr committed:
docs: update README with FCM app build identifier note and ensure proper formatting in configuration files

Sarthak-btr committed:
refactor: remove Notifee integration and replace with react-native-push-notification for notification handling

Sarthak-btr committed:
refactor: replace AsyncStorage with local storage utility for notification settings management

Sarthak-btr committed:
test: enhance Jest setup with comprehensive mocks for AsyncStorage, MMKV, Firebase, and notification services

Sarthak-btr committed:
refactor: remove AsyncStorage dependency from package.json

Sarthak-btr committed:
refactor: remove AsyncStorage and Notifee mocks from Jest setup

jjalan commented on android/app/google-services.json:
```diff
@@ -0,0 +1,86 @@
+{
+  "project_info": {
```
In this [PR](https://github.com/jalantechnologies/react-native-template/pull/165), I see some of these values are pulled from CI env.

We want one way to store these values.

It might be useful to update this PR to ensure this PR also pulls the value from env as well

Sarthak-btr replied:
The Google Services plugin expects a valid JSON file only. That is why it is not able to receive the values from the env file. It cannot process template files containing environment variable.

jjalan replied:
Should this be auto generated then when app is built using values from env?

What is the best way to address the concern highlighted in original comment.

Sarthak-btr replied:
Updated the Firebase configuration using the env variables and remove the google-service.json and Google.Service-Info.plist files. - [commit](https://github.com/jalantechnologies/react-native-template/pull/157/commits/7027d2ab4510e45b9195111d18b541a42b7ee3c6)

jjalan commented on ios/Boilerplate/GoogleService-Info.plist:
Can the file name be case insensitive?

Sarthak-btr replied:
No, the filename GoogleService-Info.plistis not case-insensitive in practice for iOS apps.

Sarthak-btr replied:
Updated the Firebase configuration using the env variables and remove the google-service.json and Google.Service-Info.plist files. - [commit](https://github.com/jalantechnologies/react-native-template/pull/157/commits/7027d2ab4510e45b9195111d18b541a42b7ee3c6)

jjalan commented on jest.setup.js:
```diff
@@ -9,6 +9,67 @@ jest.mock('react-native/Libraries/Animated/NativeAnimatedHelper');
 
 jest.mock('@datadog/mobile-react-native');
 
+// Mock MMKV
```
Why are these changes needed?

Also, see - [How to write better code comments](https://dev.to/adammc331/todo-write-a-better-comment-4c8c)

Sarthak-btr replied:
These changes are necessary because, without them, the following error occurs during the checks (compose: test):

![image](https://github.com/user-attachments/assets/880d6cda-86ca-4e7c-ab8a-fa187ba4715b)

jjalan commented on package.json:
```diff
@@ -38,6 +41,7 @@
     "react-native-error-boundary": "^1.2.3",
     "react-native-gesture-handler": "^2.12.0",
     "react-native-mmkv": "^2.10.1",
+    "react-native-push-notification": "^8.1.1",
```
Is this by RN team or 3rd party? We want to avoid dependency on 3rd party

Sarthak-btr replied:
remove both react-native-push-notification and @react-native-community/push-notification-ios third party dependencies
Now I am using only 
 "@react-native-firebase/app": "^22.2.1",
 "@react-native-firebase/messaging": "^22.2.1",
    
Now it is looking like below image for **foreground** (for background it is same):  
Should I add the custom notification drop down which looks like: (next Image)
![image](https://github.com/user-attachments/assets/7d1b87a4-f208-4fd3-9fea-6d4501c914ad)


Should I add the custom notification drop down which looks like:

![image](https://github.com/user-attachments/assets/f7a34597-0a67-4c82-a007-fee9f0a62e55)

Sarthak-btr committed:
refactor: remove both react-native-push-notification and @react-native-community/push-notification-ios third party dependencies

Sarthak-btr committed:
refactor: remove comprehensive mocks for MMKV, react-native-config, react-native-safe-area-context, react-native-gesture-handler, Firebase modules, and notification services from Jest setup

Sarthak-btr committed:
refactor: remove @react-native-firebase from transformIgnorePatterns in Jest config

Sarthak-btr committed:
fix: add missing newline at end of jest.config.js and jest.setup.js

Sarthak-btr committed:
refactor: update Jest configuration to include @react-native-firebase in transformIgnorePatterns and add mock implementations for Firebase messaging

Sarthak-btr committed:
feat: update the ios files

jjalan commented on ios/Boilerplate/AppDelegate.mm:
file name and folder name should be case insensitive

jjalan commented on ios/Boilerplate/GoogleService-Info.plist:
filename should be case insensitive.

Also can we use a different folder name? When we create a new app using this template, having a folder name boilerplate seems odd and rename a folder would cause unnecessary changes.

Sarthak-btr replied:
Should I replace it with AppStarter or ReactNativeBase?
Renaming the Boilerplate folder can break Xcode project references, pod paths, and build scripts, requiring manual updates in .xcodeproj, .xcworkspace, Podfile, and possibly CI/CD configs.

jjalan commented on ios/Boilerplate/GoogleService-Info.plist:
```diff
@@ -0,0 +1,30 @@
+<plist version="1.0">
+<dict>
+	<key>API_KEY</key>
+	<string>AIzaSyDn7Pa-H3Yjt6_Aegnkz11MlzGpV69FyV0</string>
```
Does this and other values in this file needs to change per project? I do not see it in README file.

Sarthak-btr replied:
Updated the Firebase configuration using the env variables and remove the google-service.json and Google.Service-Info.plist files. - [commit](https://github.com/jalantechnologies/react-native-template/pull/157/commits/7027d2ab4510e45b9195111d18b541a42b7ee3c6)

jjalan commented on jest.setup.js:
```diff
@@ -24,3 +24,43 @@ jest.mock('react-i18next', () => ({
   },
 }));
+
+// Provide mock implementations for Firebase App and Messaging modules to simulate Firebase behavior in tests
```
Why do we need these changes?

Sarthak-btr replied:
Note: added @react-native-firebase to avoid the conflicts in the CI/CD pipeline failing checks (compose: test)
![image](https://github.com/user-attachments/assets/161624d8-2245-41db-8dc8-57d876c497a7)

jjalan commented on jest.setup.js:
```diff
@@ -24,3 +24,43 @@ jest.mock('react-i18next', () => ({
   },
 }));
+
+// Provide mock implementations for Firebase App and Messaging modules to simulate Firebase behavior in tests
```
See - [How to write better code comments](https://dev.to/adammc331/todo-write-a-better-comment-4c8c)

jjalan commented on src/contexts/notification-context.tsx:
```diff
@@ -0,0 +1,164 @@
+  importance: 'default',
+};
+
+/**
```
[How to write better code comments](https://dev.to/adammc331/todo-write-a-better-comment-4c8c)

jjalan commented on src/contexts/notification-context.tsx:
```diff
@@ -0,0 +1,164 @@
+ * Do Not Disturb mode on Android devices for urgent communications.
+ */
+const DEFAULT_NOTIFICATION_SETTINGS: NotificationSettings = {
+  masterToggle: true,
```
Here and other places, order them alphabetically.

jjalan commented on src/services/firebase-messageing.service.ts:
```diff
@@ -0,0 +1,180 @@
+   */
+  readonly requestUserPermission = async (): Promise<void> => {
+    try {
+      if (Platform.OS === 'ios') {
```
use enum instead of string comparison

jjalan commented on src/services/firebase-messageing.service.ts:
```diff
@@ -0,0 +1,180 @@
+    try {
+      if (Platform.OS === 'ios') {
+        const authStatus = await this.messaging.requestPermission();
+        const enabled = authStatus === 1 || authStatus === 2; //
```
This code is not readable as its not clear what 1 and 2 stands for

jjalan commented on src/types/notification.types.ts:
```diff
@@ -0,0 +1,21 @@
+  enabled: boolean;
+  sound: boolean;
+  vibration: boolean;
+  importance: 'default' | 'high' | 'low';
```
create enum

Sarthak-btr committed:
feat: integrate Firebase configuration and setup for Android and iOS

Sarthak-btr committed:
refactor: remove redundant comments from notification and Firebase services for clarity

Sarthak-btr committed:
refactor: update notification types and improve Firebase messaging service for better clarity and functionality

Sarthak-btr committed:
refactor: reorganize AndroidManifest.xml structure and remove unused permissions for improved clarity

Sarthak-btr committed:
refactor: update Firebase configuration for Android and iOS, enhancing clarity and structure

Sarthak-btr committed:
refactor: update Firebase configuration instructions and improve code structure for better clarity and maintainability

Sarthak-btr committed:
refactor: fix indentation and formatting in MainApplication.java for improved readability

Sarthak-btr committed:
refactor: fix formatting and remove unnecessary hyphen in MainApplication.java for improved readability

Sarthak-btr committed:
refactor: fix formatting in MainApplication.java for improved readability

Sarthak-btr committed:
refactor: fix formatting by removing unnecessary whitespace in MainApplication.java

Sarthak-btr committed:
refactor: fix indentation in AppDelegate.mm for improved readability

Sarthak-btr committed:
refactor: update notification handling logic and improve logging in NotificationService

Patel-world commented on android/app/src/main/AndroidManifest.xml:
```diff
@@ -11,7 +11,7 @@
     >
       <activity
-        android:name=".SplashActivity"
+        android:name=".MainApplication"
```
MainApplication under activity tag?

Confirm that .MainApplication is not being incorrectly registered as an Activity, and correct the tag.

Sarthak-btr committed:
refactor: update activity name from MainApplication to SplashActivity in AndroidManifest.xml

Sarthak-btr committed:
Merge branch 'main' into sarthak/feat/implement-push-notifications-support

Sarthak-btr committed:
chore: bump version to 0.0.4 in package.json

Sarthak-btr committed:
chore: add Android keystore configuration to .env

Sarthak-btr committed:
Merge branch 'main' into sarthak/feat/implement-push-notifications-support
